File: package/node/src/core/session.ts
import type { TokenResponse } from "../types";

interface SessionStorage {
  getItem(key: string): string | null;
  setItem(key: string, value: string): void;
  removeItem(key: string): void;
}

class MemoryStorage implements SessionStorage {
  private store = new Map<string, string>();

  getItem(key: string): string | null {
    return this.store.get(key) || null;
  }

  setItem(key: string, value: string): void {
    this.store.set(key, value);
  }

  removeItem(key: string): void {
    this.store.delete(key);
  }
}

function getStorage(): SessionStorage {
  if (typeof globalThis !== "undefined" && "localStorage" in globalThis) {
    return (globalThis as unknown as { localStorage: SessionStorage }).localStorage;
  }
  return new MemoryStorage();
}

class SessionManager {
  private storage: SessionStorage;
  private readonly ACCESS_TOKEN_KEY = "aether_access_token";
  private readonly REFRESH_TOKEN_KEY = "aether_refresh_token";
  private readonly EXPIRES_AT_KEY = "aether_expires_at";

  constructor() {
    this.storage = getStorage();
  }

  setTokens(tokens: TokenResponse): void {
    this.storage.setItem(this.ACCESS_TOKEN_KEY, tokens.accessToken);
    this.storage.setItem(this.REFRESH_TOKEN_KEY, tokens.refreshToken);
    const expiresAt = Date.now() + tokens.expiresIn * 1000;
    this.storage.setItem(this.EXPIRES_AT_KEY, expiresAt.toString());
  }

  getAccessToken(): string | null {
    return this.storage.getItem(this.ACCESS_TOKEN_KEY);
  }

  getRefreshToken(): string | null {
    return this.storage.getItem(this.REFRESH_TOKEN_KEY);
  }

  getExpiresAt(): number | null {
    const expiresAt = this.storage.getItem(this.EXPIRES_AT_KEY);
    return expiresAt ? parseInt(expiresAt, 10) : null;
  }

  isAuthenticated(): boolean {
    const expiresAt = this.getExpiresAt();
    if (!expiresAt) return false;
    return Date.now() < expiresAt;
  }

  isTokenRefreshing(): boolean {
    return false;
  }

  setAccessToken(token: string, expiresIn: number): void {
    this.storage.setItem(this.ACCESS_TOKEN_KEY, token);
    const expiresAt = Date.now() + expiresIn * 1000;
    this.storage.setItem(this.EXPIRES_AT_KEY, expiresAt.toString());
  }

  setToken(token: string): void {
    this.storage.setItem(this.ACCESS_TOKEN_KEY, token);
    const expiresAt = Date.now() + 3600 * 1000;
    this.storage.setItem(this.EXPIRES_AT_KEY, expiresAt.toString());
  }

  clear(): void {
    this.storage.removeItem(this.ACCESS_TOKEN_KEY);
    this.storage.removeItem(this.REFRESH_TOKEN_KEY);
    this.storage.removeItem(this.EXPIRES_AT_KEY);
  }
}

export { SessionManager };

File: package/rust/src/core/session.rs
use std::sync::{Arc, Mutex};

use crate::types::TokenResponse;\nuse chrono::Utc;

#[derive(Clone, Debug, Default)]
pub struct SessionManager {
    access_token: Option<String>,
    refresh_token: Option<String>,
    expires_at: Option<i64>,
}

impl SessionManager {
    pub fn new(access_token: Option<String>) -> Self {
        Self {
            access_token,
            refresh_token: None,
            expires_at: None,
        }
    }

    pub fn set_tokens(&mut self, tokens: TokenResponse) {
        self.access_token = Some(tokens.access_token);
        self.refresh_token = Some(tokens.refresh_token);
        self.expires_at = Some(chrono::Utc::now().timestamp() + tokens.expires_in as i64);
    }

    pub fn get_access_token(&self) -> Option<String> {
        self.access_token.clone()
    }

    pub fn get_refresh_token(&self) -> Option<String> {
        self.refresh_token.clone()
    }

    pub fn get_expires_at(&self) -> Option<i64> {
        self.expires_at
    }

    pub fn is_authenticated(&self) -> bool {
        match self.expires_at {
            Some(ts) => chrono::Utc::now().timestamp() < ts,
            None => false,
        }
    }

    pub fn clear(&mut self) {
        self.access_token = None;
        self.refresh_token = None;
        self.expires_at = None;
    }
}

